Output parsing enhancement

alygin opened the pull request:
I've improved parsing of output:

1. Errors and Warnings are now fully matched with those of rustc.
2. Notes and Helps are added to the corresponding Errors and Warnings as Info blocks of `trace`. They can also exist as standalone messages.
3. Additional lines with useful information are parsed and added to the appropriet message blocks.
4. Panics are parsed along with their Stack backtraces (which are added as elements of their `trace`).
5. There's only one entry point for output parsing and it fully supports both standard human-readable output as well as JSON format.

So, all the compiler and program messages are now neatly arranged, no information is lost in both standard and JSON modes. Works well with clippy.

Please, check, if everything is ok. Any feedback on the code, usability and anything else is highly appreciated.

oli-obk commented:
regular parsing looks good. I'm slightly confused about json, things like `mismatched type` don't actually show the expected/found part, even though it's definitely in the json. Is that a new issue or was that there before?

oli-obk commented on lib/cargo.js:
```diff
@@ -82,65 +82,131 @@ export function provideBuilder() {
+              } else {
+                sub = {
+                  message: message,
+                  type: 'info'
```
if you attach file/line info to notes, you get a nice clickable link inside the message, please add that. Some errors are hard to diagnose from the popup alone without this info

alygin replied:
Totally agree, I'll add them.

oli-obk commented on lib/cargo.js:
```diff
@@ -82,65 +82,131 @@ export function provideBuilder() {
+              } else {
+                // Just a description in the current block. Only add it when in submessage
+                // because Linter does the job for high-level messages.
+                if (sub !== null) {
```
this could be an else if to remove some nesting

alygin replied:
The reason I have this additional nesting level is that otherwise I would have to either make two matches on the previous level (one for error/warning/note/help, another for panic) which might slightly affect performance or use `else if ((match = regex.match(...))` that would work, but is quite error-prone for future maintenance (and probably for static analysis tools), so I decided to leave it the way I did.
But I can make it more flat, that's not a problem at all.

oli-obk replied:
oh, I just meant the most inner `if(match) { .. } else { if (sub !== null) {.. } }`, But it's not a blocker if you think there's any reason to do it this way.

alygin replied:
I see now. You're right, it certainly should be simplified.

alygin commented:
@oli-obk, I left JSON parsing intact, just moved the loop on output lines one method higher. So, if there where some issues before, they are still there and it's unlikely that I introduced new ones. But I'll have a look at the problem you described since I'm fixing things there anyway.

alygin commented:
I've made a trivial fix in json parsing, the message you expected is displayed now. But it seems that the json structure has changed since the parser was written and now they're out of sync. I think we need a separate issue for that because some preliminary research must be done to do the work correctly.

Links are also added to submessages and a minor code refactoring performed. Please, check it out.

oli-obk commented:
> The notion of transwarp beaming is like trying to hit a bullet with a smaller bullet whilst wearing a blindfold, riding a horse. 

Since the json flag has a nice warning message, let's ignore whatever happens there xD

This looks good to me. Do you have anything additional or are we good to merge?

alygin commented:
Absolutely )
I've done with what I was aiming and think we could merge if there're no objections.
